### PR TITLE
Microsoft: Recurrence exceptions and cancellations

### PR DESCRIPTION
**Pull request by squeaky-pl**
This provides support for edits and deletions inside recurrences. This works by plugging the logic I developed before that fetches recurrence instances and then compares them against what would recurrence look if there were no edits to created cancellations and exceptions.

### CODE REVIEW

File: inbox/events/microsoft/parse.py [hunk 1]
```diff
@@ -287,72 +288,54 @@ def convert_msgraph_patterned_recurrence_to_ical_rrule(
     )
 
 
-class SynthetizedCanceledoccurrence:
```
Comment (squeaky-pl) [inline inbox/events/microsoft/parse.py:290]: I've changed this to just work as MsGraphEvent, even though those are not events coming from Ms Graph API. It simplified implementation and caused less problems with MyPy.

File: inbox/events/microsoft/events_provider.py [hunk 2]
```diff
@@ -12,19 +12,31 @@
...
 CALENDAR_LIST_WEBHOOK_URL = URL_PREFIX + "/w/microsoft/calendar_list_update/{}"
 EVENTS_LIST_WEBHOOK_URL = URL_PREFIX + "/w/microsoft/calendar_update/{}"
 
 
+# Microsoft Graph supports infinite and finite recursions.
+# By default recurring events are created 6 months into the
+# future but you can override it of course in the UI.
+# To prevent infinite or very long looping when searching
+# for exceptions and cancellations we need to establish a limit.
+MAX_RECURRING_EVENT_DURATION = datetime.timedelta(days=365)
```
Comment (cschams) [inline inbox/events/microsoft/events_provider.py:37]: Maybe `MAX_RECURRING_EVENT_WINDOW`? Duration made me think it was how long the single event could be.

File: inbox/events/microsoft/parse.py [hunk 3]
```diff
@@ -287,72 +288,54 @@ def convert_msgraph_patterned_recurrence_to_ical_rrule(
...
+def synthetize_canceled_occurrence(
```
Comment (cschams) [inline inbox/events/microsoft/parse.py:291]: spelling: synthesize

File: inbox/events/microsoft/parse.py [hunk 4]
```diff
@@ -602,7 +604,7 @@ def parse_event(
         recurrence = None
         start_tz = None
 
-    if event["type"] == "exception":
+    if event["type"] in ["exception", "synthetizedCancellation"]:
```
Comment (cschams) [inline inbox/events/microsoft/parse.py:607]: Here and elsewhere: "synthesizedCancellation"

**Review by cschams — APPROVED** [review]
Just fix the few naming issues and this looks good to me.

File: inbox/events/microsoft/parse.py (hunk 3 above)
Comment (squeaky-pl) [reply inbox/events/microsoft/parse.py:291]: LOL, I've been spelling this wrong all my life.

**Pull request merged**